### ISSUE TITLE
Consider using Beacon API for sending payload on `unload` event

### ISSUE THREAD

**Issue by jahtalab** [elastic/apm-agent-rum-js#195]
Currently the RUM agent sends transactions to the APM server in short intervals (500ms by default). Even with this short interval, there is still a possibility of losing data if the page is unloaded before we have a chance to send the queued transactions. Since we should not block the unload event for sending the APM data, we can use the [Beacon API](https://developer.mozilla.org/en-US/docs/Web/API/Beacon_API) for this purpose.

Please note that the Beacon API is still experimental and is not supported in all of the browsers we currently target.

**Comment by matthewarmand** [comment]
Any plans to implement this? This is the only way to reliably send transactions on certain events, notably `window.unload`, which is important for us as APM users.

> Please note that the Beacon API is still experimental and is not supported in all of the browsers we currently target.

Since this issue was first written, it seems the [browser support](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon#browser_compatibility) for the Beacon API has matured. I'm not sure which browsers APM targets, but it seems like this would be much more stable and worth it to implement now than it may have been 2 years ago.

**Comment by bryantcj52** [comment]
I tried to send a beacon request to the APM server manually but to send the correct JSON content-type header, a Blob is required which then expects all CORs headers to be set which is failing because the apm-server doesn't send back Access-Control-Accept-Credentials: true.

```Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Credentials' header in the response is '' which must be 'true' when the request's credentials mode is 'include'.```

**Comment by jahtalab** [comment]
Update: We're going to use the `visibilitychange` event to send the beacon instead of `unload`, to avoid the issues mentioned on [beacon documentation]( https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon)

We will also truncate any running transaction and send them on the beacon

**Comment by mixolydian251** [comment]
Depending on the API flexibility you need, the `fetch` API with `keepalive: true` may be another alternative to `sendBeacon`

We ran into some limitations with sendBeacon recently. We needed to set custom headers as well as had a preference for json payload. After some trial and error we found that fetch worked for our use case.

**Comment by David-Development** [comment]
@jahtalab Is there an ETA on when this will be implemented? I tried to work around this by using the `apmRequest` interceptor and setting `keepalive` to true just like @mixolydian251 suggested but the queue is not flushed before the page is gone. 

I found [this](https://github.com/elastic/apm-agent-js-core/pull/5/files) pull request in your old repo where someone added the unload handler already.
